ingester/ingester_lifecycle.go

```diff
@@ -0,0 +1,419 @@
+func (i *Ingester) ChangeState(state ring.IngesterState) error {
+	err := make(chan error)
+	i.actorChan <- func() {
+		err <- i.changeState(state)
+	}
+	return <-err
+}
+
+// ClaimTokensFor takes all the tokans for the supplied ingester and assigns them to this ingester.
+func (i *Ingester) ClaimTokensFor(ingesterID string) error {
+	err := make(chan error)
+
+	i.actorChan <- func() {
+		var tokens []uint32
+
+		claimTokens := func(in interface{}) (out interface{}, retry bool, err error) {
+			ringDesc, ok := in.(*ring.Desc)
+			if !ok || ringDesc == nil {
+				return nil, false, fmt.Errorf("Cannot claim tokens in an empty ring")
+			}
+
+			tokens = ringDesc.ClaimTokens(ingesterID, i.id)
+			return ringDesc, true, nil
+		}
+
+		if err := i.consul.CAS(ring.ConsulKey, claimTokens); err != nil {
+			log.Errorf("Failed to write to consul, sleeping: %v", err)
+		}
+
+		i.tokens = tokens
+		err <- nil
+	}
+
+	return <-err
+}
+
+// Shutdown stops the ingester.  It will:
+// - send chunks to another ingester, if it can.
+// - otherwise, flush chunks to the chunk store.
+// - remote config from Consul.
+// - block until we'll successfully shutdown.
+func (i *Ingester) Shutdown() {
+	// This will prevent us accepting any more samples
+	i.stopLock.Lock()
+	i.stopped = true
+	i.stopLock.Unlock()
+
+	// closing i.quit triggers loop() to exit, which in turn will trigger
+	// the removal of our tokens etc
+	close(i.quit)
+
+	i.done.Wait()
+}
+
+func (i *Ingester) loop() {
+	defer func() {
+		log.Infof("Ingester.loop() exited gracefully")
+		i.done.Done()
+	}()
+
+	// First, see if we exist in the cluster, update our state to match if we do,
+	// and add ourselves (without tokens) if we don't.
+	if err := i.initRing(); err != nil {
+		log.Fatalf("Failed to join consul: %v", err)
+	}
+
+	// We do various period tasks
+	autoJoinAfter := time.After(i.cfg.JoinAfter)
+
+	heartbeatTicker := time.NewTicker(i.cfg.HeartbeatPeriod)
+	defer heartbeatTicker.Stop()
+
+	flushTicker := time.NewTicker(i.cfg.FlushCheckPeriod)
+	defer flushTicker.Stop()
+
+	rateUpdateTicker := time.NewTicker(i.cfg.userStatesConfig.RateUpdatePeriod)
+	defer rateUpdateTicker.Stop()
+
+loop:
+	for {
+		select {
+		case <-autoJoinAfter:
+			// Will only fire once, after auto join timeout.  If we haven't entered "JOINING" state,
+			// then pick some tokens and enter ACTIVE state.
+			if i.state == ring.PENDING {
+				log.Infof("Auto-joining cluster after timout.")
+				if err := i.autoJoin(); err != nil {
+					log.Fatalf("failed to pick tokens in consul: %v", err)
+				}
+			}
+
+		case <-heartbeatTicker.C:
+			consulHeartbeats.Inc()
+			if err := i.updateConsul(); err != nil {
+				log.Errorf("Failed to write to consul, sleeping: %v", err)
+			}
+
+		case <-flushTicker.C:
+			i.sweepUsers(false)
+
+		case <-rateUpdateTicker.C:
+			i.userStates.updateRates()
+
+		case f := <-i.actorChan:
+			f()
+
+		case <-i.quit:
+			break loop
+		}
+	}
+
+	// Shutdown order is important!
```

Comment:
Removed comment - it used to be important when this code was in main.go, now its obvious.